src/Symfony/Component/HttpClient/EventSource.php

```diff
@@ -0,0 +1,121 @@
+    const ON_MESSAGE = 'eventsource:onmessage';
+    const ON_OPEN = 'eventsource:onopen';
+    const ON_ERROR = 'eventsource:onerror';
+
+    public $lastEventId = '';
+    public $readyState = self::CLOSED;
+
+    private $client;
+    private $reconnectionTime = 30;
+    private $shouldDisconnect = false;
+
+    public function __construct(HttpClientInterface $client = null)
+    {
+        $this->client = $client ?: HttpClient::create($configuration['default_options'] ?? []);
+    }
+
+    public function connect(string $url)
+    {
+        $parts = parse_url($url);
+        if (!isset($parts['scheme'], $parts['host']) || !\in_array($parts['scheme'], ['http', 'https'])) {
+            throw new InvalidArgumentException();
+        }
+        $this->loop($url);
+    }
+
+    public function disconnect()
+    {
+        $this->shouldDisconnect = true;
+    }
+
+    private function doRequest(string $url)
+    {
+        $headers = [
+            'Accept' => 'text/event-stream',
+            'Cache-Control' => 'no-cache',
+        ];
+
+        if ('' !== $this->lastEventId) {
+            $headers['Last-Event-ID'] = $this->lastEventId;
+        }
+
+        $this->readyState = self::CONNECTING;
+
+        return $this->client->request('GET', $url, ['headers' => $headers]);
+    }
+
+    private function loop(string $url)
+    {
+        $response = $this->doRequest($url);
+
+        if (200 !== $response->getStatusCode()) {
+            throw new ClientException($response);
+            $this->readyState = self::CLOSED;
+            $response->cancel();
+
+            return;
+        }
+
+        $this->readyState = self::OPEN;
+        $this->dispatch(new MessageEvent(), self::ON_OPEN);
+
+        $buffer = '';
+        foreach ($this->client->stream($response, $this->reconnectionTime) as $chunk) {
+            if ($this->shouldDisconnect || $chunk->isTimeout()) {
+                $this->readyState = self::CLOSED;
+                if (false === $this->shouldDisconnect) {
+                    $this->loop($url);
+                }
+
+                return;
+            }
+
+            // We have to gather our own Buffer that may be longer then a chunk
+            $buffer .= $chunk->getContent();
+
+            if (!$buffer) {
+                // connection closed
+                return;
+            }
+
+            // If the line starts with a U+003A COLON character (':') ignore the line
+            if (':' === $buffer[0]) {
+                $buffer = '';
+                continue;
+            }
+
+            // Process if we got new line ending
+            $end = substr($buffer, -1);
+            if ("\n" !== $end || "\r" !== $end) {
+                continue;
+            }
+
+            $messageEvent = MessageEvent::parse($buffer);
+            $this->lastEventId = $messageEvent->lastEventId;
+            $buffer = '';
+
+            if ('retry' === $messageEvent->eventName && \is_int($time = (int) $messageEvent->data)) {
```

Comment:
`retry` isn't an event name, it's a prop that can be added to any event (similar to `eventName`).